articles/142_idl.md

```diff
@@ -35,3 +35,53 @@ Further it describes how the interfaces are used to generate code in C, C++ and
 <div class="alert alert-warning" markdown="1">
   <b>TODO:</b> All of this too
 </div>
+
+### Type Mapping
+
+The next table defines how IDL types are mapped to the following programming languages:
+
+* C11 (or higher)
+* C++11 (or higher)
+* Python 3 (or higher)
+
+| IDL type    | C type      | C++ type    | Python type          |
+| --------    | ----------- | ----------- | -------------------- |
+| float       | float       | float       | float                |
+| double      | double      | double      | float                |
+| long double | long double | long double | float                |
+| char        | signed char | char        | str with length 1    |
+| wchar       | char16\_t   | char16\_t   | str with length 1    |
+| bool        | \_Bool      | bool        | bool                 |
+| octet       | uint8\_t    | uint8\_t    | bytes with length 1  |
+| int8        | int8\_t     | int8\_t     | int                  |
+| uint8       | uint8\_t    | uint8\_t    | int                  |
+| int16       | int16\_t    | int16\_t    | int                  |
+| uint16      | uint16\_t   | uint16\_t   | int                  |
+| int32       | int32\_t    | int32\_t    | int                  |
+| uint32      | uint32\_t   | uint32\_t   | int                  |
+| int64       | int64\_t    | int64\_t    | int                  |
+| uint64      | uint64\_t   | uint64\_t   | int                  |
+
+The next table defines how IDL templated and constructed types are mapped to the programming languages (unless specified otherwise below).
+
+| IDL type       | C type                 | C++ type            | Python type |
+| -------------- | ---------------------- | ------------------- | ----------- |
+| T[N]           | T[N]                   | std::array<T, N>    | list        |
+| sequence<T>    | struct {size\_t, T * } | std::vector<T>      | list        |
+| sequence<T, N> | struct {size\_t, T * } | custom\_class<T, N> | list        |
+| string         | char *                 | std::string         | str         |
+| string<N>      | char *                 | std::string         | str         |
+| wstring        | char16\_t *            | std::u16string      | str         |
+| wstring<N>     | char16\_t *            | std::u16string      | str         |
```

Comment:
I was wondering about the different handling of bounded sequences compares to bounded (w)strings in C++. For the first we go through the hassle to provide our own data structure (to enforce the upper bound) for the other two we don't (atm).

Maybe we should use the same approach homogeneously - it could be either of the two, The effort and "customness" of types we define vs. the delayed check until the data structure is being passed to the lower levels.

Reply:
yeah I was thinking the same for the C part, currently in C we have custom data structure for all the dynamic types (bounded and unbounded). AFAICT in C++ we have a custom container for bounded vectors but nothing for strings ATM.

In the delayed check approach, when / how would the user be notified that the data doesnt fit?

Reply:
>  when / how would the user be notified that the data doesnt fit?

The `rclpy_publish` function converts the Python object into a ROS C message. If any of the constraints are violated it can signaled through an exception.

Optionally we can offer a validation function which the user could invoke explicitly.

Reply:
In Python I would expect the user to not be allowed to place data bigger than the message specification at the Python level already (what we currently do with custom setters).

My question was about the C language that doesn't provide exceptions. If we map to a `char(16_t) *` I am not sure how we would restrict the memory allocation or report the error.

But your point still holds: it can be done when publish is called rather than when filling the message datastructure.

Reply:
> In Python I would expect the user to not be allowed to place data bigger than the message specification at the Python level already (what we currently do with custom setters).

That is true but only covers the case where a field is being "assigned". But e.g. a list is mutable and new items can be added after the initial assignment with the property of the class being able to intercept (assuming we don't mock all method of the list instance).

> My question was about the C language that doesn't provide exceptions. If we map to a char(16_t) * I am not sure how we would restrict the memory allocation or report the error.

In a future `rclc` client library the `publish` call can perform the validation the same way as in Python. While the `publish` in `rcl` could also check the constraints I would rather not do that and rely in the client library on top to do that. All of these functions have return code to signal error already (if exceptions are not available / used).

Reply:
> While the publish in rcl could also check the constraints I would rather not do that and rely in the client library on top to do that.

Just adding it looks like either client libraries or the rmw implementations must check the constraints because `rcl_publish` sees the message as a [`void *`](https://github.com/ros2/rcl/blob/0a6d9186e8fbd133308d0ad4bc3b75d61078b04d/rcl/src/rcl/publisher.c#L234).

> In Python I would expect the user to not be allowed to place data bigger than the message specification at the Python level already (what we currently do with custom setters).

I think ros2/ros2#509 is a hint that ros2 should consider only checking the validity of the message once. From the perspective of ROS 2 as a library the message constraints don't matter until the message is used. From the perspective of a developer writing code constraint checking when a field is set is nice because it errors closer to the bug, but from the perspective of a user running the code it can be a significant performance hit for little gain when most of the time everything is working fine.